### ISSUE TITLE
New rule: Generic type constraints must be on their own line

### ISSUE THREAD

**Issue by sharwell** [DotNetAnalyzers/StyleCopAnalyzers#1176]
**Category:** Readability
**Title:** Generic type constraints must be on own line
**Description:** For maximum readability, the constraints clause for a generic type parameter should be listed on a line of code by themselves. The following code would produce this violation:

```csharp
private void Method<T>() where T : class
{
}
```

To fix a violation of this rule, place the constraints for each generic type on their own line:

```csharp
private void Method<T>()
    where T : class
{
}
```

:memo: For types with more than one generic type parameter, each parameter would need to appear on a line by itself. For example, the following shows the correct way to write three constraints:

```csharp
private void Method<T1, T2, T3>()
    where T1 : class
    where T2 : class
    where T3 : class
{
}
```

**Comment by hvanbakel** [comment]
What would happen with a second or third type constraint? All on a line of their own?

**Comment by sharwell** [comment]
@hvanbakel That is correct. I'll update the proposal to include an example.

**Comment by hvanbakel** [comment]
That's relevant since you can have dependencies between them, e.g.
```CSharp
public class Test<A, B> 
   where A : class,
   where B : A
{

}
```

**Comment by JohanLarsson** [comment]
This is nice, R# has a similar feature.

**Comment by vweijsters** [comment]
:+1: 

All examples are about methods, but I assume the same holds for type declarations?

**Comment by sharwell** [comment]
> All examples are about methods, but I assume the same holds for type declarations?

That is correct :smile:

**Comment by nbarbettini** [comment]
Working on a diagnostic and code fix for this one.

**Comment by sharwell** [comment]
If you visit [this page](https://github.com/DotNetAnalyzers) and accept the invitation, I can "officially" assign the issue to you. :smile:

**Comment by nbarbettini** [comment]
@sharwell done!

**Comment by sharwell** [comment]
Assigned!

**Comment by nbarbettini** [comment]
@sharwell is there a SA(xxxx) number for this one yet? How do we assign diagnostic IDs to new rules?

**Comment by pdelvo** [comment]
We just use the next free id for the category. So for this rule this would be SA1127.

**Comment by nbarbettini** [comment]
@pdelvo sweet, thanks.

**Comment by nbarbettini** [comment]
The diagnostic for this one was pretty easy. Done [here](https://github.com/nbarbettini/StyleCopAnalyzers/blob/90e1a320e3e814e216b6712df2d925a51f6f58b9/StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1127GenericTypeConstraintsMustBeOnOwnLine.cs).

The code fix is fairly easy too, I think but so far all my attempts have the dumbz because I can't figure out how to get rid of a leading space (that is actually the previous token's trailing trivia). I'm probably missing something simple and I will keep hammering it so I can learn.
